tracing-benchmarks/src/jmh/java/com/palantir/tracing/TracingBenchmark.java

```diff
@@ -0,0 +1,152 @@
+public class TracingBenchmark {
+
+    private static final Callable<Object> NOOP_CALLABLE = () -> null;
+
+    private static final Callable<Object> undecidedCallable = Tracers.wrapWithNewTrace("test",
+            Observability.UNDECIDED, NOOP_CALLABLE);
+
+    private static final Callable<Object> sampledCallable = Tracers.wrapWithNewTrace("test",
+            Observability.SAMPLE, NOOP_CALLABLE);
+
+    private static final Callable<Object> unsampledCallable = Tracers.wrapWithNewTrace("test",
+            Observability.DO_NOT_SAMPLE, NOOP_CALLABLE);
+
+    @Setup
+    public final void before() {
+        Tracer.setSampler(new RandomSampler(0.1f));
+    }
+
+    @TearDown
+    public final void after() {
+        Tracer.setSampler(AlwaysSampler.INSTANCE);
+    }
+
+    @Benchmark
+    public static Object raw() throws Exception {
+        return NOOP_CALLABLE.call();
+    }
+
+    @Benchmark
+    public static Object undecidedCallable() throws Exception {
+        return undecidedCallable.call();
+    }
+
+    @Benchmark
+    public static Object sampledCallable() throws Exception {
+        return sampledCallable.call();
+    }
+
+    @Benchmark
+    public static Object unsampledCallable() throws Exception {
+        return unsampledCallable.call();
+    }
+
+    @Benchmark
+    public static void unsampledTracerSpan(Blackhole blackhole) throws Exception {
+        Tracer.initTrace(Observability.DO_NOT_SAMPLE, Tracers.randomId());
+        for (int i = 0; i < 100; i++) {
+            Callable<Object> callable = () -> {
+                Tracer.startSpan("span");
+                try {
+                    return NOOP_CALLABLE.call();
+                } finally {
+                    Tracer.fastCompleteSpan();
+                }
+            };
+            blackhole.consume(callable.call());
+        }
+    }
+
+    @Benchmark
+    public static void undecidedTracerSpan(Blackhole blackhole) throws Exception {
+        Tracer.initTrace(Observability.UNDECIDED, Tracers.randomId());
+        for (int i = 0; i < 100; i++) {
+            Callable<Object> callable = () -> {
+                Tracer.startSpan("span");
+                try {
+                    return NOOP_CALLABLE.call();
+                } finally {
+                    Tracer.fastCompleteSpan();
+                }
+            };
+            blackhole.consume(callable.call());
+        }
+    }
+
+    @Benchmark
+    public static void sampledTracerSpan(Blackhole blackhole) throws Exception {
+        Tracer.initTrace(Observability.SAMPLE, Tracers.randomId());
+        for (int i = 0; i < 100; i++) {
+            Callable<Object> callable = () -> {
+                Tracer.startSpan("span");
+                try {
+                    return NOOP_CALLABLE.call();
+                } finally {
+                    Tracer.fastCompleteSpan();
+                }
+            };
+            blackhole.consume(callable.call());
```

Comment:
We don't need to blackhole the null result from our callable, however we should blackhole the output, which in this case is a series of `Span` objects:
```java
Tracer.subscribe("jmh", blackhole::consume);
```

Reply:
@carterkozak  thanks for CR, does it matter that every call of `Tracer.startSpan("span");` starts a span with the same name?

Reply:
Spans with the same name are totally fine, we have them in practice via recursive methods ;-)

Reply:
good example. do you want each span subscribed by `blackhole::consume`?

Reply:
We just need to set the subscriber at the beginning when we get the blackhole instance. Ideally we could configure that globally instead of for each benchmark method, but I’m not sure how. Note that tracing observer registration is a static per jvm, so concurrent measurements will reset the observer

Reply:
We just need to set the subscriber at the beginning when we get the blackhole instance. Ideally we could configure that globally instead of for each benchmark method, but I’m not sure how. Note that tracing observer registration is a static per jvm, so concurrent measurements will reset the observer

Reply:
@carterkozak it's a static map, i think i can give different names for different benchmark if that's your concern?

Reply:
if we run a benchmark with >1 threads, the black holes will be clobbered.

We want to implement a state object (bean annotated with `@org.openjdk.jmh.annotations.State(Scope.Benchmark)`) where we can request a Blackhole instance in a `@org.openjdk.jmh.annotations.Setup` method and subscribe there, paired with a `@TearDown` to remove the observer. Then we don't need to worry about subscribing in the invocation (and we can properly run multi-threaded benchmarks).

Reply:
👍

Reply:
do you have an example of request a `Blackhole` instance in `setup` method? I guess if the whole benchmark class is annotated with `@State(Scope.Benchmark)`. The same instance of Blackhole gets passed into those benchmarked methods. but i don't have proof.

Reply:
nope they are not the same instance.

Reply:
@carterkozak  can i use a `UUID` as the observer name? so that each invocation has different observer.

Reply:
No, you'll end up writing each span to N observers instead of the case we want to benchmark with a single observer.

Reply:
I'd start with something along these lines:

```java
@Fork(1)
@Threads(1)
@State(Scope.Benchmark)
@Warmup(iterations = 3, time = 3)
@Measurement(iterations = 3, time = 3)
public class SimpleBenchmark {

    @Setup
    public void setup(Blackhole blackhole) {
        Tracer.subscribe("jmh", blackhole::consume);
        Tracer.setSampler(AlwaysSampler.INSTANCE);
    }

    @TearDown
    public void tearDown() {
        Tracer.unsubscribe("jmh");
    }

    @Benchmark
    public void nestedSpans() {
        for (int i = 0; i < 100; i++) {
            Tracer.startSpan("span");
        }
        for (int i = 0; i < 100; i++) {
            Tracer.fastCompleteSpan();
        }
    }
}
```

Reply:
The sampler can be parameterized so that we can write one benchmark and get results for both the sampled and unsampled cases using parameterized values in state, see the internal class I linked in a DM.

Reply:
cool 👍